java/org/apache/catalina/realm/DataSourceRealm.java

```diff
@@ -539,6 +612,162 @@ private boolean isRoleStoreDefined() {
     }
 
 
+    /**
+     * Return the specified user's requested user attributes as a map.
+     * 
+     * @param dbConnection The database connection to be used
+     * @param username User name for which to return user attributes
+     * 
+     * @return a map containing the specified user's requested user attributes
+     */
+    protected Map<String, Object> getUserAttributesMap(Connection dbConnection, String username) {
+
+        String preparedAttributes = getUserAttributesStatement(dbConnection);
+        if (preparedAttributes == null || preparedAttributes == USER_ATTRIBUTES_NONE_REQUESTED) {
```

Comment:
Why not `preparedAttributes.isEmpty()`?

Reply:
Ah, `isEmpty()` is clearly better. Didn't have that method in mind (only thought of `preparedAttributes.length() == 0` and wanted something more _speaking_). Would you keep the `USER_ATTRIBUTES_NONE_REQUESTED` constant anyway?

Reply:
Personally, I see very little use unless there is a very special semantics. I hardly see any.

Reply:
Guten Morgen :-)
Yes, that wwere my doubts, too. So, I changed that to `isEmpty()`. My question was whether I should keep the constant when setting `preparedAttributes`to the empty string by returning it method `getUserAttributesStatement`. Here `return USER_ATTRIBUTES_NONE_REQUESTED;`is likely more descriptive than `return "";`.

Reply:
Why not return `null`? Is is necessary for string concat?

Reply:
With DataSourceRealm, parsing and validating is done lazily, so `null` means _still uninitialized_. When not using a value different from `null` for _empty/none required_, I end with constantly parsing and validating in method `getUserAttributesStatement`. I could have uses another boolean `userAttributesNoneRequested`, however I like that  _3-state value approach_ some more.

Reply:
Yes, that's a problem. I agree. Since I cannot provide anything better I accept this.